interfaces/mount/backend_test.go

```diff
@@ -167,7 +167,44 @@ func (s *backendSuite) TestSetupSetsupWithoutDir(c *C) {
 
 	// Ensure that backend.Setup() creates the required dir on demand
 	os.Remove(dirs.SnapMountPolicyDir)
-	s.InstallSnap(c, interfaces.ConfinementOptions{}, mockSnapYaml, 0)
+	s.InstallSnap(c, interfaces.ConfinementOptions{}, "snap-name", mockSnapYaml, 0)
+}
+
+func (s *backendSuite) TestParallelInstanceSetup(c *C) {
+	old := dirs.SnapDataDir
+	defer func() {
+		dirs.SnapDataDir = old
+	}()
+	dirs.SnapDataDir = "/var/snap"
+	snapEntry := osutil.MountEntry{Name: "/snap/snap-name_instance", Dir: "/snap/snap-name", Type: "none", Options: []string{"rbind", osutil.XSnapdOriginOvername()}, DumpFrequency: 0, CheckPassNumber: 0}
+	dataEntry := osutil.MountEntry{Name: "/var/snap/snap-name_instance", Dir: "/var/snap/snap-name", Type: "none", Options: []string{"rbind", osutil.XSnapdOriginOvername()}, DumpFrequency: 0, CheckPassNumber: 0}
+	fsEntry1 := osutil.MountEntry{Name: "/src-1", Dir: "/dst-1", Type: "none", Options: []string{"bind", "ro"}, DumpFrequency: 0, CheckPassNumber: 0}
+	fsEntry2 := osutil.MountEntry{Name: "/src-2", Dir: "/dst-2", Type: "none", Options: []string{"bind", "ro"}, DumpFrequency: 0, CheckPassNumber: 0}
+	fsEntry3 := osutil.MountEntry{Name: "/src-3", Dir: "/dst-3", Type: "none", Options: []string{"bind", "ro"}, DumpFrequency: 0, CheckPassNumber: 0}
+
+	// Give the plug a permanent effect
+	s.Iface.MountPermanentPlugCallback = func(spec *mount.Specification, plug *snap.PlugInfo) error {
+		if err := spec.AddMountEntry(fsEntry1); err != nil {
+			return err
+		}
+		return spec.AddUserMountEntry(fsEntry3)
+	}
+	// Give the slot a permanent effect
+	s.iface2.MountPermanentSlotCallback = func(spec *mount.Specification, slot *snap.SlotInfo) error {
+		return spec.AddMountEntry(fsEntry2)
+	}
+
+	// confinement options are irrelevant to this security backend
+	s.InstallSnap(c, interfaces.ConfinementOptions{}, "snap-name_instance", mockSnapYaml, 0)
+
+	// Check that snap fstab file contains parallel instance setup and data from interfaces
+	expected := strings.Join([]string{snapEntry.String(), dataEntry.String(), fsEntry2.String(), fsEntry1.String()}, "\n") + "\n"
+	fn := filepath.Join(dirs.SnapMountPolicyDir, "snap.snap-name_instance.fstab")
+	c.Check(fn, testutil.FileEquals, expected)
+
+	// Check that the user-fstab file was written with user mount only
+	fn = filepath.Join(dirs.SnapMountPolicyDir, "snap.snap-name_instance.user-fstab")
+	c.Check(fn, testutil.FileEquals, fsEntry3.String()+"\n")
```

Comment:
Nitpick: I think this test would be easier to scan if fsEntry{1,2,3} were split into fsEntry{1,2} and userFsEntry1. Right now to understand the association of particular numbers to particular parts of the profile is subtle.